libs/nest/config/src/lib/ConfigurationLoader.ts

```diff
@@ -0,0 +1,162 @@
+import { ConfigDefinition, Configuration, EnvLoader } from './types'
+import { ServerSideFeatureClient } from '@island.is/feature-flags'
+import { logger } from '@island.is/logging'
+import { InvalidConfiguration } from './InvalidConfiguration'
+import { ConfigurationError } from './ConfigurationError'
+import { ZodError } from 'zod'
+
+type Issue = 'missing' | 'could not be parsed as JSON'
+
+export class ConfigurationLoader<T> implements EnvLoader {
+  private isDev = process.env.NODE_ENV !== 'production'
+  private issues: Record<string, Issue> = {}
+
+  constructor(private definition: ConfigDefinition<T>) {}
+
+  load(): Configuration<T> {
+    if (
+      this.definition.serverSideFeature &&
+      !ServerSideFeatureClient.isOn(this.definition.serverSideFeature)
+    ) {
+      logger.info({
+        category: 'ConfigModule',
+        message: `Ignored configuration for ${this.definition.name}. Server-side feature flag missing: ${this.definition.serverSideFeature}`,
+      })
+      return new InvalidConfiguration(
+        `Unable to read configuration for ${this.definition.name}. Server-side feature flag missing: ${this.definition.serverSideFeature}`,
+      ) as Configuration<T>
+    }
+
+    try {
+      const result = this.definition.load(this)
+
+      // Loader did not crash, but there might still be issues we need to deal with.
+      return this.handleResult(result)
+    } catch (err) {
+      // Loader crashed. If there are reported issues, throw those instead.
+      return this.handleResult(undefined, err)
+    }
+  }
+
+  private handleResult(result?: T, error?: Error) {
+    const issues = Object.entries(this.issues)
+    const errorMessage = this.formatIssues(issues)
+    const hasParseIssue = issues.find(
+      (issue) => issue[1] === 'could not be parsed as JSON',
+    )
+    const hasMissingIssue = !hasParseIssue && issues.length > 0
+
+    // Always throw if there are any parse issues.
+    if (hasParseIssue) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Only throw missing issue in production
+    if (hasMissingIssue && !this.isDev) {
+      throw new ConfigurationError(errorMessage)
+    }
+
+    // Re-throw other unexpected errors from loader.
+    if (error) {
+      throw error
+    }
```

Comment:
there's a bunch of code around handling the issue and issue types both here and in `formatIssues()` etc.. i wonder if the issue thing would be more clear if it had its own class definition

eg.
```
class Issues {
  issues: Record<string, Issue> = {}
  hasMissingIssue() {}
  hasParseIssue() {}
  formatIssues () {}
}
```

might also read a lot clearer

Reply:
Good suggestion, will consider.